src/helpers/controllers/dynamic-grid-sizes.ts

```diff
@@ -0,0 +1,243 @@
+import { Signal } from "@lit-labs/preact-signals";
+import { VerificationGridComponent } from "../../components/verification-grid/verification-grid";
+import { Size } from "../../models/rendering";
+import { Pixel, UnitInterval } from "../../models/unitConverters";
+
+export interface GridShape {
+  rows: number;
+  columns: number;
+}
+
+export class DynamicGridSizeController<Container extends HTMLElement> {
+  public constructor(container: Container, targetElement: VerificationGridComponent, isOverlapping: Signal<boolean>) {
+    this.targetElement = targetElement;
+    this.isOverlapping = isOverlapping;
+
+    const resizeObserver = new ResizeObserver((size: ResizeObserverEntry[]) => {
+      if (!this.target) return;
+
+      // because we know that this resize observer will only ever be used with
+      // one element, we can safely assume that the first element in the array
+      // is the container element
+      const container = size[0];
+      this.handleResize(container);
+    });
+    resizeObserver.observe(container);
+
+    // because each tile advertises if they are overlapping, we debounce the
+    // overlapping signal so that if one grid size causes multiple tiles to
+    // overlap (and the signal is set multiple times), we only have to handle
+    // the intersection once
+    let debounceTimeout: number | undefined;
+    this.isOverlapping.subscribe((value: boolean) => {
+      if (debounceTimeout) {
+        clearTimeout(debounceTimeout);
+      }
+
+      debounceTimeout = window.setTimeout(() => {
+        if (value) {
+          this.handleIntersection();
+        }
+      }, 5);
+    });
+  }
```

Comment:
is this debounce even needed? I thought signals were meant to deduplicate input?